sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala

```diff
@@ -0,0 +1,120 @@
+                        cosmosPatchConfigs: CosmosPatchConfigs) {
+ private val log = LoggerHelper.getLogger(diagnosticsConfig, this.getClass)
+
+ private val TimestampAttributeName = "_ts"
+ private val IdAttributeName = "id"
+ private val ETagAttributeName = "_etag"
+ private val SelfAttributeName = "_self"
+ private val ResourceIdAttributeName = "_rid"
+ private val AttachmentsAttributeName = "_attachments"
+
+ private val systemProperties = List(
+  TimestampAttributeName,
+  ETagAttributeName,
+  SelfAttributeName,
+  ResourceIdAttributeName,
+  AttachmentsAttributeName)
+
+ def isAllowedProperty(path: String, partitionKeyDefinition: PartitionKeyDefinition): Boolean = {
+  assertNotNullOrEmpty(path, "path")
+  assertNotNull(partitionKeyDefinition, "partitionKeyDefinition")
+
+  // There are some properties are immutable, these kind properties include:
+  // 1. System properties : _ts, _rid, _etag
+  // 2. id, and partitionKeyPath
+  if ((path.startsWith("/") && !systemProperties.contains(path.substring(1)))
+   && IdAttributeName != path
+   && !StringUtils.join(partitionKeyDefinition.getPaths, "").contains(path)) {
+   true
+  } else {
+   false
+  }
+ }
+
+ def addOperationConditionally(cosmosPatchOperations: CosmosPatchOperations,
+                               columnConfig: CosmosPatchColumnConfig,
+                               objectNode: ObjectNode,
+                               condition: Boolean,
+                               message: String): Unit = {
+  if (condition) {
+   columnConfig.operationType match {
+    case CosmosPatchOperationTypes.Add => cosmosPatchOperations.add(columnConfig.mappingPath, objectNode.get(columnConfig.columnName))
+    case CosmosPatchOperationTypes.Set => cosmosPatchOperations.set(columnConfig.mappingPath, objectNode.get(columnConfig.columnName))
+    case CosmosPatchOperationTypes.Replace => cosmosPatchOperations.replace(columnConfig.mappingPath, objectNode.get(columnConfig.columnName))
+    case CosmosPatchOperationTypes.Remove => cosmosPatchOperations.remove(columnConfig.mappingPath)
+    case CosmosPatchOperationTypes.Increment => addIncrementPatchOperation(cosmosPatchOperations, columnConfig, objectNode.get(columnConfig.columnName))
+    case _ => throw new IllegalArgumentException(s"Patch operation type ${columnConfig.operationType} is not supported")
+   }
+  } else {
+   log.logDebug(
+    s" The operation will not be added due to condition checking failed," +
+     s" columnName: ${columnConfig.columnName}, opType: ${columnConfig.operationType}, message: $message")
+  }
+ }
+
+ def createCosmosPatchOperations(partitionKeyDefinition: PartitionKeyDefinition,
+                                 objectNode: ObjectNode): CosmosPatchOperations = {
+
+  val cosmosPatchOperations = CosmosPatchOperations.create()
+
+  cosmosPatchConfigs.columnConfigsMap.values.foreach(columnConfig => {
+   if (isAllowedProperty(columnConfig.mappingPath, partitionKeyDefinition)) {
```

Comment:
There is a possibility that there will no patch operation included. In this case, when we send an empty request to backend, we are going to get 304 (Not modified) back.

Reply:
I would actually throw an error in that case

Reply:
Would hardly ever be intentional

Reply:
but I see in spark, the customer can define includeNonNull or NonDefault for the serialize settings, so I feel it could increase the chance when this check will be true. 

should we be more relaxing the restriction now?

Reply:
Discussed offline, agreed that it is better to start with more restrict rules to give customer a clear signal to check the data or the configs. In the future, if it is a common scenarios, we can change to relax the check here by adding a extra config in patch to allow empty operations.